internal/services/tesla_fleet_api_service.go

```diff
@@ -164,12 +178,81 @@ func (t *teslaFleetAPIService) GetAvailableCommands() *UserDeviceAPIIntegrations
 	}
 }
 
-// performTeslaGetRequest a helper function for making http requests, it adds a timeout context and parses error response
-func (t *teslaFleetAPIService) performTeslaGetRequest(ctx context.Context, url, token string) (*http.Response, error) {
+// VirtualTokenConnectionStatus Checks whether vehicles can accept Tesla commands protocol for the partner's public key
+func (t *teslaFleetAPIService) VirtualTokenConnectionStatus(ctx context.Context, token, region, vin string) (bool, error) {
+	baseURL := fmt.Sprintf(t.Settings.TeslaFleetURL, region)
+	url := fmt.Sprintf("%s/api/1/vehicles/fleet_status", baseURL)
+
+	jsonBody := []byte(fmt.Sprintf(`{"vins": ["%s"]}`, vin))
+	body := bytes.NewReader(jsonBody)
+
+	resp, err := t.performRequest(ctx, url, token, http.MethodPost, body)
+	if err != nil {
+		return false, fmt.Errorf("could not fetch vehicles for user: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	var v VirtualTokenConnectionStatusResponse
+	bd, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("could not verify connection status %w", err)
+	}
+
+	err = json.Unmarshal(bd, &v)
+	if err != nil {
+		return false, fmt.Errorf("error occurred decoding connection status %w", err)
+	}
+
+	isConnected := slices.Contains(v.Response.KeyPairedVins, vin)
+
+	return isConnected, nil
+}
+
+func (t *teslaFleetAPIService) RefreshToken(ctx context.Context, refreshToken string) (*TeslaAuthCodeResponse, error) {
+	reqs := struct {
+		GrantType    string `json:"grant_type"`
+		ClientID     string `json:"client_id"`
+		RefreshToken string `json:"refresh_token"`
+	}{
+		GrantType:    "refresh_token",
+		ClientID:     t.Settings.TeslaClientID,
+		RefreshToken: refreshToken,
+	}
+
+	reqb, err := json.Marshal(reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", "https://auth.tesla.com/oauth2/v3/token", bytes.NewBuffer(reqb))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/json")
```

Comment:
I think the body is to be form-encoded. See [Tesla](https://developer.tesla.com/docs/fleet-api#third-party-token) and [the spec](https://datatracker.ietf.org/doc/html/rfc6749#page-47).